### PR TITLE
Manually down-integrate python JSON struct support from internal code base.

### PR DESCRIPTION
**Pull request by calbach**
**NOTE:** Includes some manual reverts from the internal code base to avoid interfering with more recent GitHub commits.

Ignores new generated subdirectory: `python/google/protobuf/util/`

### CODE REVIEW

File: python/google/protobuf/internal/json_format_test.py [hunk 1]
```diff
@@ -404,16 +596,13 @@ def testParseBadEnumValue(self):
 
   def testParseBadIdentifer(self):
     self.CheckError('{int32Value: 1}',
-                    (r'Failed to load JSON: Expecting property name'
-                     r'( enclosed in double quotes)?: line 1'))
```
Comment (anandolee) [inline python/google/protobuf/internal/json_format_test.py]: This should not change. 
I will do an up-integrate first

Comment (anandolee) [inline python/google/protobuf/internal/json_format_test.py]: Er... we may not want to do up-integrate right now which will cause a lot of work.
Can you manually revert this kind of changes? There are 3 small changes you may want to take case of for this file:
https://github.com/google/protobuf/commits/master/python/google/protobuf/internal/json_format_test.py

File: python/google/protobuf/internal/well_known_types_test.py [hunk 2]
```diff
@@ -36,18 +36,18 @@
 
 from datetime import datetime
 
-try:
-  import unittest2 as unittest
-except ImportError:
-  import unittest
-
```
Comment (anandolee) [inline python/google/protobuf/internal/well_known_types_test.py]: Revert it:
https://github.com/calbach/protobuf/commits/python-json-struct/python/google/protobuf/internal/well_known_types_test.py

File: python/google/protobuf/internal/well_known_types_test.py [hunk 3]
```diff
@@ -36,18 +36,18 @@
...
+from google.protobuf import any_pb2
 from google.protobuf import duration_pb2
 from google.protobuf import field_mask_pb2
+from google.protobuf import struct_pb2
 from google.protobuf import timestamp_pb2
+import unittest
```
Comment (anandolee) [inline python/google/protobuf/internal/well_known_types_test.py]: Revert

**Comment by calbach** [comment]
Thanks for catching that. I believe I've made all the necessary changes except one. The one thing I'm not sure how to resolve is the change to `_RegularMessageToJsonObject` which was changed [here](https://github.com/calbach/protobuf/commit/2850a98275b1397c511973e207a16ea704ffb18f#diff-aae725250c924f43781bccd0505ed91fL117). Is that change still relevant? Tests appear to be passing locally for me.

**Comment by calbach** [comment]
Scratch that, I decided to revert back to what's in GitHub since it seems this was necessary.

**Comment by anandolee** [comment]
Please change the the PR name to "Manually Down-integrate ...", because this is not our regular down-integrate and we did some manually revert in it. Others LGTM

**Comment by calbach** [comment]
Done. Thanks for the review.

File: python/google/protobuf/json_format.py [hunk 4]
```diff
@@ -200,6 +204,77 @@ def _FieldToJsonObject(
...
+  """Converts ListValue message according to Proto3 JSON Specification."""
+  return [_ValueMessageToJsonObject(value)
+          for value in message.values]
+
+
+def _StructMessageToJsonObject(message, unused_including_default=False):
+  """Converts Struct message according to Proto3 JSON Specification."""
+  fields = message.fields
+  return {key: _ValueMessageToJsonObject(fields[key])
+          for key in fields}
```
Comment (anandolee) [inline python/google/protobuf/json_format.py]: Looks like python2.6 does not support this:
https://travis-ci.org/google/protobuf/jobs/105725474#L1751. 

Feel free to change it or delete python2.6 test (add back in another PR): https://github.com/calbach/protobuf/blob/2850a98275b1397c511973e207a16ea704ffb18f/travis.sh#L134

Comment (calbach) [inline python/google/protobuf/json_format.py]: Attempted to fix. Couldn't test locally so we'll see if travis succeeds this time..

**Comment by calbach** [comment]
Oops, I failed to rename the commit message on the previous push... now it's fixed. My python 2.6 fix is finally about to run it seems: https://travis-ci.org/google/protobuf/builds/105756178

**Comment by pgrosu** [comment]
@calbach  Thank you for putting this together.  One little thing that might save time, is that you can run the tests manually on a local machine if you want, since waiting for Travis might take a while.  For instance, I have Python 2.7.9, and I took this pull request and ran the Python conformance test like this specifically for version 2.7:

```Bash
$ wget --quiet https://codeload.github.com/calbach/protobuf/zip/5477f8cdbab0103beba17fc90ae8730835ea427f
$ mv 5477f8cdbab0103beba17fc90ae8730835ea427f protobuf.zip
$ unzip protobuf.zip
$ cd protobuf-5477f8cdbab0103beba17fc90ae8730835ea427f
$ ./autogen.sh
$ ./configure
$ make -j2
$ wget --quiet https://pypi.python.org/packages/source/p/py/py-1.4.31.tar.gz
$ tar xzvf py-1.4.31.tar.gz
$ cd py-1.4.31
$ python setup.py install --user 
$ cd ..
$ wget --quiet https://pypi.python.org/packages/source/t/tox/tox-2.3.1.tar.gz
$ tar zxvf tox-2.3.1.tar.gz
$ cd tox-2.3.1
$ python setup.py install --user 
$ cd ..
$ cd python
$ export LD_LIBRARY_PATH=../src/.libs # for Linux
$ python -mtox -e 'py{27}-python'
```
The end of the output was the following:

```Bash
  JsonInput.Struct.JsonOutput
  JsonInput.Struct.ProtobufOutput
  JsonInput.ValueAcceptBool.JsonOutput
  JsonInput.ValueAcceptBool.ProtobufOutput
  JsonInput.ValueAcceptFloat.JsonOutput
  JsonInput.ValueAcceptFloat.ProtobufOutput
  JsonInput.ValueAcceptInteger.JsonOutput
  JsonInput.ValueAcceptInteger.ProtobufOutput
  JsonInput.ValueAcceptList.JsonOutput
  JsonInput.ValueAcceptList.ProtobufOutput
  JsonInput.ValueAcceptObject.JsonOutput
  JsonInput.ValueAcceptObject.ProtobufOutput
  JsonInput.ValueAcceptString.JsonOutput
  JsonInput.ValueAcceptString.ProtobufOutput

CONFORMANCE SUITE PASSED: 421 successes, 0 skipped, 54 expected failures, 0 unexpected failures.

conformance_python: received EOF from test runner after 475 tests, exiting
______________________________________________________________ summary ______________________________________________________________
  py27-python: commands succeeded
  congratulations :)
```
Hope it helps and have great weekend!
~p

**Pull request merged**